api/filters/replacement/replacement.go

```diff
@@ -0,0 +1,153 @@
+	for _, fp := range target.FieldPaths {
+		fieldPath := strings.Split(fp, ".")
+		// TODO (#3492): Add tests for map keys in the fieldPath (e.g. .spec.containers[name=nginx])
+		t, err := node.Pipe(yaml.Lookup(fieldPath...))
+		if err != nil {
+			return err
+		}
+		if t != nil {
+			// TODO (#3492): Use the field options to refine interpretation of the field
+			t.SetYNode(value.YNode())
+		}
+	}
+	return nil
+}
+
+func getReplacement(nodes []*yaml.RNode, r *types.Replacement) (*yaml.RNode, error) {
+	source, err := selectSourceNode(nodes, r.Source)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.Source.FieldPath == "" {
+		r.Source.FieldPath = types.DefaultFieldPath
+	}
+	fieldPath := strings.Split(r.Source.FieldPath, ".")
+
+	// TODO (#3492): Add tests for map keys in the fieldPath (e.g. .spec.containers[name=nginx])
+	rn, err := source.Pipe(yaml.Lookup(fieldPath...))
+	if err != nil {
+		return nil, err
+	}
+	// TODO (#3492): Use the field options to refine interpretation of the field
+	return rn, nil
+}
+
+// selectSourceNode finds the node that matches the selector, returning
+// an error if multiple or none are found
+func selectSourceNode(nodes []*yaml.RNode, selector *types.SourceSelector) (*yaml.RNode, error) {
+	var matches []*yaml.RNode
+	for _, n := range nodes {
+		include, err := matchSource(n, selector)
+		if err != nil {
+			return nil, err
+		}
+		if include {
+			if len(matches) > 0 {
+				return nil, fmt.Errorf("more than one match for source %v", selector)
+			}
+			matches = append(matches, n)
+		}
+	}
+	if len(matches) == 0 {
+		return nil, fmt.Errorf("found no matches for source %v", selector)
+	}
+	return matches[0], nil
+}
+
+// selectTargetNodes finds the nodes that matches the TargetSelector
+// TODO (#3492): Don't include matches listed in the `reject` field
+func selectTargetNodes(nodes []*yaml.RNode, objRef *types.TargetSelector) ([]*yaml.RNode, error) {
+	var result []*yaml.RNode
+	for _, n := range nodes {
+		include, err := matchTarget(n, objRef.Select)
+
+		if err != nil {
+			return nil, err
+		}
+		if include {
+			result = append(result, n)
+		}
+	}
+	return result, nil
+}
+
+func matchSource(n *yaml.RNode, s *types.SourceSelector) (bool, error) {
+	ns, err := n.GetNamespace()
+	if err != nil {
+		return false, err
+	}
+	apiVersion := yaml.GetValue(n.Field(yaml.APIVersionField).Value)
+	return (n.GetName() == s.Name || s.Name == "") &&
+		(n.GetKind() == s.Kind || s.Kind == "") &&
+		(ns == s.Namespace || s.Namespace == "") &&
+		(apiVersion == s.ApiVersion() || s.ApiVersion() == ""), nil
+}
+
+func matchTarget(n *yaml.RNode, s *types.Selector) (bool, error) {
+	ns, err := n.GetNamespace()
+	if err != nil {
+		return false, err
+	}
+	apiVersion := yaml.GetValue(n.Field(yaml.APIVersionField).Value)
+	return (n.GetName() == s.Name || s.Name == "") &&
+		(n.GetKind() == s.Kind || s.Kind == "") &&
+		(ns == s.Namespace || s.Namespace == "") &&
+		(apiVersion == s.ApiVersion() || s.ApiVersion() == ""), nil
+}
```

Comment:
I'm working on making the change (shouldn't take long) but just adding as a note that this will change a lot of tests because it changes how the selector object is constructed